### DISCUSSION TITLE
Finding the command has same effect with Xnormal Dilation in Photoshop

### DISCUSSION THREAD

**Discussion by kelliray90** [ImageMagick/ImageMagick#6496] [Help]
### ImageMagick version

7.1.0-58 Q16-HDR

### Operating system, version and so on

windows 10

### Description

I'm finding the command that has same effect like Xnormal Dilation in Photoshop. 

The most similar command I found : `magick orig.png -morphology Dilate Octagon:1 morphology-dilate.png`

There're 2 difference things between the above command and Xnormal Dilation. I wanna know how to let -morphology work as the same with Xnormal Dilation. 
Download the image to test here: [dilation.zip](https://github.com/ImageMagick/ImageMagick/files/12082972/dilation.zip)

The first: 2 different areas are smeared even though they don't touch each other.
![Untitled-5](https://github.com/ImageMagick/ImageMagick/assets/28614684/07be71a1-ea53-48c0-88bf-c6642486519d)

The second: Xnormal only dilates in transparent areas while -morphology dilates even when not transparent. How to make -morphology only dilates in tranparent areas?

**Comment by fmw42** [comment]
I see no Xnormal Dilation nor any dilation in Photoshop.  Where is that configured?  Please always give more information such that we can reproduce your effect.  Also define Xnormal Dilation or say how it is different from normal dilation.

**Reply by fmw42** [reply]
From what I can tell by searching Googled, Xnormal is a windows only plugin to Photoshop.   I am on a Mac. So I doubt I can be of any help.

**Comment by fmw42** [comment]
Try this

```
magick orig.png \
\( +clone -alpha extract -write mpr:alpha \) \
-alpha off -compose multiply -composite \
mpr:alpha -compose copy_opacity -composite \
-channel rgba -morphology dilate octagon:1 \
x.png
```
![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/20f69d37-d22b-4a77-b972-4005217afffa)

**Reply by kelliray90** [reply]
I tried your command and it solved the first problem, thanks for your help.
 
For the second problem, is there any things like mask for -morphology Dilate (which limit the morphology operation to a specific area of the image, in my case the specific area is the transparent area)?

**Reply by fmw42** [reply]
Process the whole image, then use a mask to blend between your input and dilated images.  -morphology does not have a built in mask option that I know about.

**Comment by snibgo** [comment]
`-morphology` does respond to a `-write-mask`. This should be black where you want the full morphology effect, white where you want no effect, or shades of gray for partial effect.

**Reply by kelliray90** [reply]
I'll try it, thank you but now, the biggest problem is the color after Dilate is brighter than orig color (both green circle and yellow box are brighter). How to solve it?

The command I used:  `magick orig-a.png ( +clone -alpha extract -write mpr:alpha ) -alpha off -compose multiply -composite mpr:alpha -compose copy_opacity -composite -channel rgba -morphology dilate octagon:3 result.png`
Image to test:
![orig-a](https://github.com/ImageMagick/ImageMagick/assets/28614684/73b65252-6ddc-4513-8784-79c053e2cfa2)

![example](https://github.com/ImageMagick/ImageMagick/assets/28614684/1301cab4-02b1-429b-b03c-04629c96e7f4)

**Comment by snibgo** [comment]
The dull orange square does not change colour.

The small green circles (almost pure green) become larger dull greenish-yellow circles. This is because you are *dilating* each channel separately, which means that where a channel value in a pixel is larger than the same channel in an adjacent pixel, the larger value spreads into the adjacent pixel.

In your case, the green from the small circle gets spread into larger circles. But the red and blue channels work in the opposite way, because the red and blue channels have larger values outside the circle, than inside the circle. So those values are spread from the outside inwards, so the colours change.

If you want the larger green circles to be the same green as the smaller ones, I suggest `DilateIntensity` instead of `Dilate`. See https://imagemagick.org/Usage/morphology/#intensity

**Comment by fmw42** [comment]
Using snibgo's suggestion, I get

![orig2](https://github.com/ImageMagick/ImageMagick/assets/13458218/b77c1b5d-fc12-41a5-a3d1-ef88c14acc6a)


```
magick orig2.png \
\( +clone -alpha extract -write mpr:alpha \) \
-alpha off -compose multiply -composite \
mpr:alpha -compose copy_opacity -composite \
-channel rgba -morphology dilateIntensity octagon:1 \
x2.png
```

![x2](https://github.com/ImageMagick/ImageMagick/assets/13458218/e9ba1a63-0081-4d56-8830-6b737f844aea)

**Reply by fmw42** [reply]
Here is a better (simpler) version of the code

```
magick orig2.png \
-background black -alpha background \
-channel rgba -morphology dilateIntensity octagon:1 \
x2b.png

```


![x2b](https://github.com/ImageMagick/ImageMagick/assets/13458218/bb18a358-e23f-4946-9423-b3d47fe98bf4)
